Services/Administration/OJS.Services.Administration.Business/Implementations/ProblemsBusinessService.cs

```diff
@@ -54,16 +58,32 @@ public class ProblemsBusinessService : IProblemsBusinessService
             this.submissionTypesData = submissionTypesData;
             this.problemGroupsBusiness = problemGroupsBusiness;
             this.contestsBusiness = contestsBusiness;
-            this.submissionsDistributorCommunication = submissionsDistributorCommunication;
+            this.submissionsCommonBusinessService = submissionsCommonBusinessService;
+            this.testData = testData;
         }
 
-        public async Task RetestById(int id)
+        public async Task RetestById(int id, bool retestBySingleTest)
         {
-            var submissions = await this.submissionsData.GetAllByProblem(id)
+            IQueryable<Submission> submissionsQueryable;
+            var problemIdFromQuery = 0;
+
+            if (retestBySingleTest)
+            {
+                problemIdFromQuery = await this.testData.GetProblemIdByTestId(id);
+
+                submissionsQueryable = this.submissionsData.GetAllByProblem(problemIdFromQuery);
+            }
+            else
+            {
+                submissionsQueryable = this.submissionsData.GetAllByProblem(id);
+            }
```

Comment:
This is veery misleading. You are essentially using the parameter `id` to mean two different things - in one case it is the `Id` of a `Problem` (`this.submissionsData.GetAllByProblem(id)`) in another case - the `Id` of a `Test` (`this.testData.GetProblemIdByTestId(id)`). The second parameter is there only to control what the firs one will mean..
It should be 2 different public methods.. internally you can still reuse logic, but for the caller it should be clear what the method does and what parameters to provide. If the method is in `ProblemsBusinessService` and is called `RetestById` - then it's a must that the `Id` is only the `Id` of the Problem.